heartbeat/monitors/monitor.go

```diff
@@ -0,0 +1,209 @@
+	// monitor we have
+	mpi, err := pluginInfo(config)
+	if err != nil {
+		return nil, err
+	}
+
+	monitorPlugin, found := registrar.get(mpi.Type)
+	if !found {
+		return nil, fmt.Errorf("monitor type %v does not exist", mpi.Type)
+	}
+
+	m := &Monitor{
+		name:           monitorPlugin.name,
+		scheduler:      scheduler,
+		jobTasks:       []*monitorTask{},
+		pipeline:       pipeline,
+		watchPollTasks: []*monitorTask{},
+		runMtx:         sync.Mutex{},
+		config:         config,
+	}
+
+	jobs, err := monitorPlugin.create(config)
+	if err != nil {
+		return nil, fmt.Errorf("job err %v", err)
+	}
+
+	m.jobTasks, err = m.makeTasks(config, jobs)
+	if err != nil {
+		return nil, err
+	}
+
+	err = m.makeWatchTasks(monitorPlugin)
+	if err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+func (m *Monitor) makeTasks(config *common.Config, jobs []Job) ([]*monitorTask, error) {
+	mtConf := monitorTaskConfig{}
+	if err := config.Unpack(&mtConf); err != nil {
+		return nil, errors.Wrap(err, "invalid config, could not unpack monitor config")
+	}
+
+	var mTasks []*monitorTask
+	for _, job := range jobs {
+		t, err := newMonitorTask(job, mtConf, m)
+		if err != nil {
+			// Failure to compile monitor processors should not crash hb or prevent progress
+			if _, ok := err.(InvalidMonitorProcessorsError); ok {
+				logp.Critical("Failed to load monitor processors: %v", err)
+				continue
+			}
+
+			return nil, err
+		}
+
+		mTasks = append(mTasks, t)
+	}
+
+	return mTasks, nil
+}
+
+func (m *Monitor) makeWatchTasks(monitorPlugin pluginBuilder) error {
+	watchCfg := watcher.DefaultWatchConfig
+	err := m.config.Unpack(&watchCfg)
+	if err != nil {
+		return err
+	}
+
+	m.watchPollTasks = nil
+
+	if len(watchCfg.Path) > 0 {
+		m.watch, err = watcher.NewFilePoller(watchCfg.Path, watchCfg.Poll, func(content []byte) {
+			dec := json.NewDecoder(bytes.NewBuffer(content))
+			for dec.More() {
+				var obj map[string]interface{}
+				err = dec.Decode(&obj)
+				if err != nil {
+					logp.Err("Failed parsing JSON object: %v", err)
+					return
+				}
+
+				cfg, err := common.NewConfigFrom(obj)
+				if err != nil {
+					logp.Err("Failed normalizing JSON input: %v", err)
+					return
+				}
+
+				merged, err := common.MergeConfigs(m.config, cfg)
+				if err != nil {
+					logp.Err("Could not merge config: %v", err)
+					return
+				}
+
+				watchJobs, err := monitorPlugin.create(merged)
+				if err != nil {
+					logp.Err("Could not create job from watch file: %v", err)
+				}
+
+				watchTasks, err := m.makeTasks(merged, watchJobs)
+				if err != nil {
+					logp.Err("Could not make task for config: %v", err)
+					return
+				}
+
+				m.watchPollTasks = append(m.watchPollTasks, watchTasks...)
+			}
+
+			m.Stop()
+			m.Start()
+		})
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *Monitor) Start() {
+	m.runMtx.Lock()
+	defer m.runMtx.Unlock()
+
+	for _, t := range m.jobTasks {
+		t.Start()
+	}
+	for _, t := range m.watchPollTasks {
```

Comment:
I'm getting confused by the naming here. I assume a pollTask is basically also a job task? Do we need to separate the list for watchPollTasks and jobTasks or could it be just one list `tasks`?

Reply:
We separate the lists because every time we reload the watch file we blow away this slice.

We could combine the lists, but then the logic is more complex, since we need to delete only watch generated tasks.

So, when watches are made obsolete we just delete `watchPollTasks`

Reply:
Got it, makes sense.